### DISCUSSION TITLE
Custom OpenAPI content type and attrs with lists of submodels bug

### DISCUSSION THREAD

**Discussion by sector119** [fastapi/fastapi#8484] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List

import yaml
from fastapi import FastAPI, HTTPException, Request
from pydantic import BaseModel, ValidationError

app = FastAPI()


class Image(BaseModel):
    name: str
    path: str


class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    images: List[Image]

    class Config:
        schema_extra = {
            "example": {
                "name": "Foo",
                "description": "A very nice Item",
                "price": 35.4,
                "images": [
                    {"name": "img1.png", "path": "/tmp/img1.png"},
                    {"name": "img2.png", "path": "/tmp/img2.png"}
                ],
            }
        }

@app.post(
    "/items",
    openapi_extra={
        "requestBody": {
            "content": {"application/x-yaml": {"schema": Item.schema()}},
            "required": True,
        },
    },
)
async def create_item(request: Request):
    raw_body = await request.body()
    try:
        data = yaml.safe_load(raw_body)
    except yaml.YAMLError:
        raise HTTPException(status_code=422, detail="Invalid YAML")
    try:
        item = Item.parse_obj(data)
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=e.errors())
    return item
```

### Description

Hello

When I use attribute that is List of Submodel I get "Could not resolve reference: Could not resolve pointer: /definitions/Image does not exist in document" error.
If I use attribute that is List of any generic type like "str" -> images: List[str] everything works as expected.
<img width="1762" alt="err" src="https://user-images.githubusercontent.com/5519953/167415439-a4a1f2fb-0ea0-4c20-8443-59862a03ebe5.png">

### Operating System

Linux

### FastAPI Version

0.75.2

### Python Version

3.10.4

**Comment by sector119** [comment]
Not a bug, but Item.schema() should be called with ref_template=REF_PREFIX + “{model}” when submodel types are used.

REF_PREFIX could be imported from fastapi.openapi.constants and it defaults to "#/components/schemas/"

That have to be mentioned in docs IMO

**Comment by leokhachatorians** [comment]
Yeah, this caused a decent amount of annoyance for me as well.

**Comment by sector119** [comment]
It doesn't help (
Image submodel definition does not move to the '/components/schemas'

Maybe someone knows how to move schemas from definitions to the /components/schemas when custom content type schema is used?

```
"definitions": {
  "Image": {
    "title": "Image",
    "type": "object",
    "properties": {
      "name": {
        "title": "Name",
        "type": "string"
      },
      "path": {
        "title": "Path",
        "type": "string"
      }
    },
    "required": [
      "name",
      "path"
    ]
  }
}
```

**Comment by robotkutya** [comment]
any update on this...?

**Comment by albertalexandrov** [comment]
Hi!
I face the same problem. Did you solve it?

**Reply by sector119** [reply]
Only by customising FastAPI.openapi method, where you have to move all your models from #/definitions/{model} to #/components/schemas/{model}
And by subclassing BaseModel and setting right default_ref_template value like :

```
import typing

from pydantic import BaseModel

default_ref_template = "#/components/schemas/{model}"


class CustomBaseModel(BaseModel):
    @classmethod
    def schema(
        cls, by_alias: bool = True, ref_template: str = default_ref_template
    ) -> dict[str, typing.Any]:
        return super().schema(by_alias=by_alias, ref_template=ref_template)

    class Config:
        orm_mode = True

```

**Reply by sector119** [reply]
You can get this repo as an example

https://github.com/sector119/custom_openapi

**Reply by albertalexandrov** [reply]
Thanks a lot)
